File: dsa_learning_series/week7/1_SUBINC.cpp

/*Longest Increasing Subsequence and Kadanes algorithm are pre-requisites to solve this problem*/
#include<iostream>
#include<vector>


using namespace std;

int main(){
	int t;
	cin >> t;
	while(t--){
		int n;
		cin >> n;
		vector<long long> v;
		for(int i=0;i<n;i++){
			long long inp;
			cin >> inp;
			v.push_back(inp);
		};
		long long dp[n];
		for(int i=0;i<n;i++){
			dp[i] = 1;
		};
		for(int i=1;i<n;i++){
			if(v[i]>=v[i-1]){
				dp[i] = max(dp[i], 1+dp[i-1]);
			}
		};
		long long sum = 0;
		for(int i=0;i<n;i++){
			sum+=dp[i];
		};
		cout << sum << endl;
	};
	return 0;
}

File: dsa_learning_series/week7/1_SUBINC.py
#longest increasing subsequence and kadanes algorithm is a pre-requisite to solve this problem

for _ in range(int(input())):
    n = int(input())
    l = list(map(int, input().split()))
    dp = [1 for _ in range(n)]

    #res->to store max length of LIS on every iteration
    res = 0

    for i in range(1, n):
        '''
        #code for LIS
        for j in range(i):
            if dp[i]>dp[j]:
                dp[i] = max(1+dp[j], dp[i])
        res = max(dp[i], res)
        '''
        #ith element has to be greater than or equal to i-1
        if l[i]>=l[i-1]:
            dp[i] = max(dp[i], 1+dp[i-1])
        
    #print(res)
    print(sum(dp))
    
